[8.19] Improvements to the gap remediation feature (#6935)
elastic/security-docs: docs/detections/rules-ui-manage.asciidoc

=====

[float]
[[bulk-fill-gaps-multiple-rules]]
=== Fill gaps for multiple rules 

From the Rules table, fill gaps for multiple rules by using the **Fill gaps** bulk action.

. Find *Detection rules (SIEM)* in the navigation menu or by using the {kibana-ref}/introduction.html#kibana-navigation-search[global search field]. 
. In the Rules table, click the **Rule Monitoring** tab, then do one of the following:
+

> ** Fill rules with unfilled or partially filled gaps - Select the appropriate rules or all rules on the page, then click **Bulk actions -> Fill gaps**.

Reviewer: ```suggestion
** Fill rules with unfilled or partially filled gaps: Select the appropriate rules or all rules on the page, then click **Bulk actions -> Fill gaps**.
```

> ** Only fill rules with unfilled gaps: In the panel above the table, click the **Only rules with unfilled gaps** filter to only show rules with unfilled gaps (rules with partially filled gaps are excluded). Select the appropriate rules or all of them, then click **Bulk actions -> Fill gaps**. 

Reviewer 2: > rules with partially filled gaps are excluded

Partially filled gaps are not necessarily excluded

We only show rules that currently have unfilled intervals (not necessarily UNFILLED status).

Only rules with unfilled gaps → Show rules that have any unfilled intervals at that moment and user **can** make action to fill it

---

### Examples

#### Unfilled gap

A rule has a gap from `12:00–13:00`:

```
unfilled_intervals: [12:00–13:00]
in_progress_intervals: []
filled_intervals: []
status: UNFILLED
````

**Only rules with unfilled gaps:** The rule **will be shown**, since it has unfilled intervals.

---

#### Gap fill in progress

User trigger "Fill gap"

```
unfilled_intervals: []
in_progress_intervals: [12:00–13:00]
filled_intervals: []
status: UNFILLED
```

Even though the status is still `UNFILLED`, there are **no unfilled intervals** right now and no more actions user can do.

**Only rules with unfilled gaps:** The rule **will not be shown**, since there’s nothing left to schedule.

---

#### Partial filled

Let's say after gap fill some intervals were filled, others were not:

```
unfilled_intervals: [12:00–12:30]
in_progress_intervals: []
filled_intervals: [12:30–13:00]
status: PARTIALLY_FILLED
```

**Only rules with unfilled gaps:** The rule **will be shown**, because unfilled intervals still exist.

---

####  Remaining gaps

The user clicks “Fill remaining gaps”, and execution is scheduled:

```
unfilled_intervals: []
in_progress_intervals: [12:00–12:30]
filled_intervals: [12:30–13:00]
status: PARTIALLY_FILLED
```

**Only rules with unfilled gaps:** The rule **will not be shown**, since the remaining gap is already in progress.

---

#### Filled

The gap fill completes:

```
unfilled_intervals: []
in_progress_intervals: []
filled_intervals: [12:00–13:00]
status: FILLED
```

**Only rules with unfilled gaps:** The rule **will not be shown**, as all gaps are filled.

Author: ```suggestion
** Only fill rules with unfilled gaps: In the panel above the table, click the **Only rules with unfilled gaps** filter to only show rules with unfilled gaps (rules with gaps that are being filled are excluded). Select the appropriate rules or all of them, then click **Bulk actions -> Fill gaps**. 
```

> ** Only fill rules with unfilled gaps - In the panel above the table, click the **Only rules with unfilled gaps** filter to only show rules with unfilled gaps (rules with partially filled gaps are excluded). Select the appropriate rules or all of them, then click **Bulk actions -> Fill gaps**. 

Reviewer: ```suggestion
** Only fill rules with unfilled gaps: In the panel above the table, click the **Only rules with unfilled gaps** filter to only show rules with unfilled gaps (rules with partially filled gaps are excluded). Select the appropriate rules or all of them, then click **Bulk actions -> Fill gaps**. 
```

+
. Specify when to start and end the manual run that will fill the gaps. 
. Click **Schedule gap fills**. The rule will manually run over unfilled gaps in the selected time range. 

After scheduling the manual run, you can track gap fill progress by checking the **Total rules with gaps:** field in the panel above the Rules table. The field displays two metrics separated by a forward slash. The metric on the left tells you the remaining number of rules with unfilled gaps. The metric on the right tells you the number of rules that are currently having their gaps filled. 

> Alternatively, you can check gap fill progress for individual rules by going to their details page, clicking the **Execution results** tab, and going to the <<gaps-table>>.

Reviewer 3: ```suggestion
Alternatively, you can check gap fill progress for individual rules by going to their details page, clicking the **Execution results** tab, then going to the <<gaps-table>>.
```
